pkg/storage/ingest/pusher.go

```diff
@@ -75,14 +80,37 @@ func (c pusherConsumer) pushRequests(ctx context.Context, reqC <-chan parsedReco
 
 		c.processingTimeSeconds.Observe(time.Since(processingStart).Seconds())
 		if err != nil {
-			level.Error(c.l).Log("msg", "failed to push write request; skipping", "err", err)
-			// TODO move distributor's isClientError to a separate package and use that here to swallow only client errors and abort on others
-			continue
+			if !isClientIngesterError(err) {
+				return fmt.Errorf("consuming record at index %d for tenant %s: %w", recordIdx, wr.tenantID, err)
+			}
+			level.Warn(c.l).Log("msg", "consuming write request", "err", err, "user", wr.tenantID)
 		}
 	}
 	return nil
 }
 
+func isClientIngesterError(err error) bool {
+	stat, ok := grpcutil.ErrorToStatus(err)
+	if !ok {
+		// This should not be reached but in case it is, fall back to assuming it's our fault.
+		return false
+	}
+	if util.IsHTTPStatusCode(stat.Code()) {
```

Comment:
I think we could require that ingesters only use gRPC errors when used with Kafka. No need to support http errors.

Reply:
yeah makes sense; changed in https://github.com/grafana/mimir/pull/6940/commits/81c49ce7bcc4d61e6354efe072f1c6c4000c0850